### PR TITLE
kubeadm config move ControlPlaneEndpoint to ClusterConfiguration

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR moves `ControlPlaneEndpoint` from the `API` config struct to `ClusterConfiguration`.

This change is required as initial step for enabling management of more than one control plane instances in kubeadm

**Which issue(s) this PR fixes** :
refs https://github.com/kubernetes/kubeadm/issues/911, refs https://github.com/kubernetes/kubeadm/issues/963

**Special notes for your reviewer**:
just an appetizer, the main dish will be the next PR...

**Release note**:
```release-note
kubeadm: ControlPlaneEndpoint was moved from the API config struct to ClusterConfiguration
```

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/sig cluster-lifecycle
/area kubeadm
/kind api-change
/kind enhancement
/assign @luxas
/assign @timothysc
/cc @chuckha @rosti @neolit123 @liztio

### CODE REVIEW

File: cmd/kubeadm/app/util/endpoint.go [hunk 1]
```diff
@@ -27,42 +27,42 @@ import (
...
 
 // GetMasterEndpoint returns a properly formatted endpoint for the control plane built according following rules:
-// - If the api.ControlPlaneEndpoint is defined, use it.
-// - if the api.ControlPlaneEndpoint is defined but without a port number, use the api.ControlPlaneEndpoint + api.BindPort is used.
-// - Otherwise, in case the api.ControlPlaneEndpoint is not defined, use the api.AdvertiseAddress + the api.BindPort.
-func GetMasterEndpoint(api *kubeadmapi.API) (string, error) {
+// - If the ControlPlaneEndpoint is defined, use it.
+// - if the ControlPlaneEndpoint is defined but without a port number, use the ControlPlaneEndpoint + api.BindPort is used.
+// - Otherwise, in case the ControlPlaneEndpoint is not defined, use the api.AdvertiseAddress + the api.BindPort.
+func GetMasterEndpoint(cfg *kubeadmapi.InitConfiguration) (string, error) {
```
Comment (rosti) [inline cmd/kubeadm/app/util/endpoint.go]: Perhaps s/InitConfiguration/ClusterConfiguration ?

**Review by rosti — APPROVED** [review]
@fabriziopandini Thanks for this PR. Looks good to me.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha2/conversion.go [hunk 2]
```diff
@@ -32,103 +32,144 @@ func Convert_v1alpha2_InitConfiguration_To_kubeadm_InitConfiguration(in *InitCon
...
 		return err
 	}
+	if err := split_v1alpha2_InitConfiguration_into_kubeadm_ClusterConfiguration(in, &out.ClusterConfiguration, s); err != nil {
+		return err
+	}
+	return nil
+}
 
+func split_v1alpha2_InitConfiguration_into_kubeadm_ClusterConfiguration(in *InitConfiguration, out *kubeadm.ClusterConfiguration, s conversion.Scope) error {
+	if err := split_v1alpha2_InitConfiguration_into_kubeadm_ComponentConfigs(in, &out.ComponentConfigs, s); err != nil {
```
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha2/conversion.go]: Do we have UT that covers the conversion tests?

Comment (timothysc) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha2/conversion.go]: Also, I'm not certain why this code is under v1alpha2 ClusterConfiguration is only v1alpha3.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 3]
```diff
@@ -385,14 +405,11 @@ func ValidateFeatureGates(featureGates map[string]bool, fldPath *field.Path) fie
 	return allErrs
 }
 
-// ValidateAPIEndpoint validates API server's endpoint
-func ValidateAPIEndpoint(c *kubeadm.API, fldPath *field.Path) field.ErrorList {
+// ValidateAPI validates API configuration
+func ValidateAPI(c *kubeadm.API, fldPath *field.Path) field.ErrorList {
```
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: why nix the `EndPoint` part?

**Review by timothysc — COMMENTED** [review]
/approve 

But please answer some of the other clarifying questions.

File: cmd/kubeadm/app/apis/kubeadm/v1alpha2/conversion.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha2/conversion.go]: > Do we have UT that covers the conversion tests?

For tests I followed Luxas approach already used for API changes during last cycle.

There is a robust and estensive integration test under `cmd/kubeadm/app/util/config` (but no UT for each function strinctly speaking); with the integration test we are doing the following roundtrips back and forth different API formats
- v1alpha2 -> internal
- v1alpha3 -> internal
- v1alpha2 -> internal -> v1alpha3
- v1alpha3 -> internal -> v1alpha3
- v1alpha2 -> default -> validate -> internal -> v1alpha3

And then we check that no data are lost, so I'm pretty confident that everything is covered.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go (hunk 3 above)
Comment (timothysc) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: sure.

**Pull request merged**